src/Hint/NoCapitalisms.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-
+    Detect uses of capitalisms
+
+    Only allow up to two consecutive capital letters in identifiers.
+
+    Identifiers containing underscores are exempted from thus rule.
+    Identifiers of FFI bindings are exempted from thus rule.
+
+<TEST>
+module SSL.Foo -- ???
+data LHsDecl
+class FOO a where -- ???
+class Foo a where getFOO -- ???
+data Foo = Bar | BAAZ -- ???
+data Foo = B_ar | BAAZ -- ???
+data Foo = Bar | B_AAZ
+data OTPToken = OTPToken -- ???
+data OTP_Token = Foo 
+sendSMS = ... -- ???
+runTLS = ... -- ???
+runTLSSocket = ... -- ???
+runTLS_Socket
+newtype TLSSettings = ... -- ???
+tlsSettings
+data CertSettings = CertSettings
+tlsServerHooks
+tlsServerDHEParams = ... -- ???
+type WarpTLSException = () -- ???
+get_SMS
+runCI
+foreign import ccall _FIREMISSLES :: IO ()
+let getSMS = x in foo --- ???
+</TEST>
+-}
+
+
+module Hint.NoCapitalisms(noCapitalismsHint) where
+
+import Hint.Type (DeclHint,remark, Severity (Ignore))
+import Data.List.Extra (nubOrd)
+import Data.List.NonEmpty (toList)
+import Data.Char
+import Data.Maybe
+
+import GHC.Types.Basic
+import GHC.Types.SourceText
+import GHC.Data.FastString
+import GHC.Hs.Decls
+import GHC.Hs.Extension
+import GHC.Hs
+import GHC.Types.SrcLoc
+
+import Language.Haskell.GhclibParserEx.GHC.Hs.Decls
+import Language.Haskell.GhclibParserEx.GHC.Utils.Outputable
+import GHC.Util
+
+noCapitalismsHint :: DeclHint
+noCapitalismsHint _ _ decl = [ remark Ignore "Avoid capitalisms" (reLoc (shorten decl))
+                             | not $ isForD decl
+                             , name <- nubOrd $ getNames decl
+                             , not $ hasUnderscore name
+                             , hasCapitalism name
+                             ]
+
+hasUnderscore :: String -> Bool
+hasUnderscore = elem '_'
+
+hasCapitalism :: String -> Bool
+hasCapitalism s = any isAllUpper (trigrams s)
+  where
+    isAllUpper = all isUpper
+    trigrams = \case
+      a:b:c:as -> [a,b,c] : trigrams (c:as)
+      _otherwise -> []
```

Comment:
Shouldn't `b` also be part of the recursive call?

```diff
-a:b:c:as -> [a,b,c] : trigrams (c:as)
+a:b:c:as -> [a,b,c] : trigrams (b:c:as)
```

Otherwise you're missing triples that exist across the boundary e.g.

```haskell
hasCapitalism "getFOO" === False -- your example, probably not what you want?
hasCapitalism "geFOO" === True   -- should be the same, in any case, right?
```